EventsourcedProcessor: write batch size exceeding partition size

Hi,

I just ran into the following exception with one of my `EventsourcedProcessor`s (a `StatefulProcessor` to be specific). During recovery the processor crashes with a write failure due to the write batch size exceeding the configured partition size (here `16000`).

```
Caused by: java.lang.IllegalArgumentException: requirement failed: write batch size (28611) must not be greater than maximum partition size (16000)
	at scala.Predef$.require(Predef.scala:224)
	at com.rbmhtechnology.eventuate.log.EventLog$.com$rbmhtechnology$eventuate$log$EventLog$$adjustSequenceNr(EventLog.scala:675)
	at com.rbmhtechnology.eventuate.log.EventLog$$anonfun$writeBatches$1.apply(EventLog.scala:558)
	at com.rbmhtechnology.eventuate.log.EventLog$$anonfun$writeBatches$1.apply(EventLog.scala:558)
	at scala.util.Try$.apply(Try.scala:192)
	at com.rbmhtechnology.eventuate.log.EventLog.writeBatches(EventLog.scala:558)
	at com.rbmhtechnology.eventuate.log.EventLog.com$rbmhtechnology$eventuate$log$EventLog$$processReplicationWrites(EventLog.scala:528)
	at com.rbmhtechnology.eventuate.log.EventLog$$anonfun$com$rbmhtechnology$eventuate$log$EventLog$$initialized$1.applyOrElse(EventLog.scala:433)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
	at com.rbmhtechnology.eventuate.log.EventLog.aroundReceive(EventLog.scala:247)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

I think the `EventsourcedProcessor`s `write()` requests a `ReplicationWrite` without respecting any batch sizes at all - not sure if that's the real issue though.

Cheers,
Gregor

--- a/eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/EventsourcedProcessor.scala
+++ b/eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/EventsourcedProcessor.scala
@@ -27,6 +27,7 @@ import com.typesafe.config.Config
 import scala.collection.immutable.Seq
 import scala.concurrent._
 import scala.concurrent.duration._
+import scala.util._
 
 private class EventsourcedProcessorSettings(config: Config) {
   val readTimeout =
@@ -73,7 +74,7 @@ object EventsourcedProcessor {
  *
  * @see [[StatefulProcessor]]
  */
-trait EventsourcedProcessor extends EventsourcedWriter[Long, Long] {
+trait EventsourcedProcessor extends EventsourcedWriter[Long, Long] with ActorLogging {
   import ReplicationProtocol._
   import context.dispatcher
 
@@ -84,7 +85,7 @@ trait EventsourcedProcessor extends EventsourcedWriter[Long, Long] {
 
   private val settings = new EventsourcedProcessorSettings(context.system.settings.config)
 
-  private var processedEvents: Vector[DurableEvent] = Vector.empty
+  private var processedEvents: Vector[Seq[DurableEvent]] = Vector.empty
   private var processingProgress: Long = 0L
 
   /**
@@ -112,19 +113,22 @@ trait EventsourcedProcessor extends EventsourcedWriter[Long, Long] {
     case payload if processEvent.isDefinedAt(payload) =>
       val currentProcessedEvents = processEvent(payload)
       if (lastSequenceNr > processingProgress)
-        processedEvents = currentProcessedEvents.map(createEvent(_, lastHandledEvent.customDestinationAggregateIds)).foldLeft(processedEvents)(_ :+ _)
+        processedEvents = processedEvents :+ currentProcessedEvents.map(createEvent(_, lastHandledEvent.customDestinationAggregateIds))
   }
 
   /**
    * Asynchronously writes processed events that have been collected since the last write together
-   * with the current processing progress.
+   * with the current processing progress. If the number of processed events since the last write
+   * is greater than the configured `eventuate.log.write-batch-size` multiple batches with a size
+   * less than or equal to `eventuate.log.write-batch-size` will be written sequentially. When
+   * splitting into multiple batches it is guaranteed that the processing result of a single event
+   * is part of the same batch i.e. it is guaranteed that the processing result of a single event
+   * is written atomically. If the processing result of a single event is larger than
+   * `eventuate.log.write-batch-size`, `write` fails with a [[EventsourcedProcessingResultTooLargeException]].
    */
   override final def write(): Future[Long] =
     if (lastSequenceNr > processingProgress) {
-      val result = targetEventLog.ask(ReplicationWrite(processedEvents, lastSequenceNr, id, VectorTime.Zero))(Timeout(writeTimeout)).flatMap {
-        case s: ReplicationWriteSuccess => Future.successful(s.storedReplicationProgress)
-        case f: ReplicationWriteFailure => Future.failed(f.cause)
-      }
+      val result = writeBatches(processedEvents, processingProgress, lastSequenceNr)
       processedEvents = Vector.empty
       result
     } else Future.successful(processingProgress)
@@ -140,17 +144,23 @@ trait EventsourcedProcessor extends EventsourcedWriter[Long, Long] {
   }
 
   /**
-   * Limits the replay batch size to `eventuate.log.write-batch-size`. Can be overridden.
-   */
-  override def replayBatchSize: Int =
-    settings.writeBatchSize
-
-  /**
    * Sets the written processing progress for this processor.
    */
   override def writeSuccess(progress: Long): Unit = {
     processingProgress = progress
     super.writeSuccess(progress)
+  }
+
+  /**
+   * Stops the processor if called with a [[EventsourcedProcessingResultTooLargeException]]
+   * otherwise delegates to `super`.
+   */
+  override def writeFailure(cause: Throwable): Unit = cause match {
+    case e: EventsourcedProcessingResultTooLargeException =>
+      log.error(cause, "shutting down processor")
+      context.stop(self)
+    case e =>
+      super.writeFailure(e)
   }
 
   /**
@@ -186,7 +196,50 @@ trait EventsourcedProcessor extends EventsourcedWriter[Long, Long] {
       customDestinationAggregateIds = customDestinationAggregateIds,
       vectorTimestamp = lastVectorTimestamp,
       processId = DurableEvent.UndefinedLogId)
+
+  private def splitBatches(events: Vector[Seq[DurableEvent]]): (Vector[Seq[DurableEvent]], Vector[Seq[DurableEvent]]) = {
+    var num = 0
+    // Ensure that a single processEvent result
+    // is not partitioned across write batches.
+    events.span { w =>
+      val size = w.size
+      if (size > settings.writeBatchSize) {
+        throw new EventsourcedProcessingResultTooLargeException(size, settings.writeBatchSize)
+      } else {
+        num += size
+        num <= settings.writeBatchSize
+      }
+    }
+  }
+
+  private def writeBatches(events: Vector[Seq[DurableEvent]], previousProgress: Long, currentProgress: Long): Future[Long] = {
+    Try(splitBatches(events)) match {
+      case Failure(e) =>
+        Future.failed(e)
+      case Success((w, r)) =>
+        (w.flatten, r) match {
+          case (Vector(), Vector()) =>
+            Future.successful(currentProgress)
+          case (batch, Vector()) =>
+            writeBatch(batch, currentProgress)
+          case (batch, tail) =>
+            writeBatch(batch, previousProgress).flatMap(_ => writeBatches(tail, previousProgress, currentProgress))
+        }
+    }
+  }
+
+  private def writeBatch(events: Seq[DurableEvent], progress: Long): Future[Long] =
+    targetEventLog.ask(ReplicationWrite(events, progress, id, VectorTime.Zero))(Timeout(writeTimeout)).flatMap {
+      case s: ReplicationWriteSuccess => Future.successful(s.storedReplicationProgress)
+      case f: ReplicationWriteFailure => Future.failed(f.cause)
+    }
 }
+
+/**
+ * Thrown if the processing result of a single event is larger than the configured `eventuate.log.write-batch-size`.
+ */
+class EventsourcedProcessingResultTooLargeException(actualSize: Int, allowedSize: Int)
+  extends RuntimeException(s"Processing result of size $actualSize is greater than eventuate.log.write-batch-size $allowedSize")
 
 /**
  * An [[EventsourcedProcessor]] that supports stateful event processing. In-memory state created from source

--- a/eventuate-core/src/test/scala/com/rbmhtechnology/eventuate/EventsourcedProcessorSpec.scala
+++ b/eventuate-core/src/test/scala/com/rbmhtechnology/eventuate/EventsourcedProcessorSpec.scala
@@ -20,6 +20,7 @@ import akka.actor._
 import akka.testkit._
 
 import com.rbmhtechnology.eventuate.EventsourcedViewSpec._
+import com.typesafe.config.ConfigFactory
 
 import org.scalatest._
 
@@ -27,6 +28,8 @@ import scala.collection.immutable.Seq
 
 object EventsourcedProcessorSpec {
   import DurableEvent._
+
+  val config = ConfigFactory.parseString("eventuate.log.write-batch-size = 10")
 
   val eventA = event("a", 1)
   val eventB = event("b", 2)
@@ -52,6 +55,12 @@ object EventsourcedProcessorSpec {
     }
 
     override val processEvent: Process = {
+      case "x" =>
+        Vector.fill(4)("x")
+      case "y" =>
+        Vector.fill(10)("y")
+      case "z" =>
+        Vector.fill(11)("z")
       case evt: String =>
         processedEvents = processedEvents :+ evt
         Seq(s"${evt}-1", s"${evt}-2")
@@ -75,7 +84,7 @@ object EventsourcedProcessorSpec {
     event.copy(emitterId = emitterIdB, processId = UndefinedLogId, localLogId = UndefinedLogId, localSequenceNr = UndefinedSequenceNr)
 }
 
-class EventsourcedProcessorSpec extends TestKit(ActorSystem("test")) with WordSpecLike with Matchers with BeforeAndAfterAll with BeforeAndAfterEach {
+class EventsourcedProcessorSpec extends TestKit(ActorSystem("test", EventsourcedProcessorSpec.config)) with WordSpecLike with Matchers with BeforeAndAfterAll with BeforeAndAfterEach {
   import EventsourcedProcessorSpec._
   import EventsourcingProtocol._
   import ReplicationProtocol._
@@ -135,13 +144,20 @@ class EventsourcedProcessorSpec extends TestKit(ActorSystem("test")) with WordSp
     else processResult(GetReplicationProgressFailure(TestException))
   }
 
-  def processWrite(progress: Long, events: Seq[DurableEvent], success: Boolean = true): Unit = {
+  def processPartialWrite(progress: Long, events: Seq[DurableEvent], success: Boolean = true): Unit = {
     trgProbe.expectMsg(ReplicationWrite(events, progress, emitterIdB, VectorTime()))
     if (success) {
       processResult(ReplicationWriteSuccess(events.size, progress, emitterIdB, VectorTime()))
-      appProbe.expectMsg(progress)
     } else {
       processResult(ReplicationWriteFailure(TestException))
+    }
+  }
+
+  def processWrite(progress: Long, events: Seq[DurableEvent], success: Boolean = true): Unit = {
+    processPartialWrite(progress, events, success)
+    if (success) {
+      appProbe.expectMsg(progress)
+    } else {
       appProbe.expectMsg(TestException)
     }
   }
@@ -255,6 +271,44 @@ class EventsourcedProcessorSpec extends TestKit(ActorSystem("test")) with WordSp
       processWrite(2, Seq(
         eventB1.copy(vectorTimestamp = timestamp(0, 1)),
         eventB2.copy(vectorTimestamp = timestamp(0, 1))))
+    }
+    "write events in multiple batches if the number of generated events during processing since the last write is greater than settings.writeBatchSize" in {
+      val actor = recoveredStatelessProcessor()
+
+      val evt1 = event("x", 1).copy(vectorTimestamp = timestamp(1, 0))
+      val evt2 = event("x", 2).copy(vectorTimestamp = timestamp(2, 0))
+      val evt3 = event("x", 3).copy(vectorTimestamp = timestamp(3, 0))
+      val evt4 = event("x", 4).copy(vectorTimestamp = timestamp(4, 0))
+
+      actor ! Written(evt1) // ensure that a write is in progress when processing the next events
+      actor ! Written(evt2)
+      actor ! Written(evt3)
+      actor ! Written(evt4)
+
+      // process first write
+      processWrite(1, Seq.fill(4)(update(evt1)))
+
+      // process remaining writes
+      processPartialWrite(1, Seq.fill(4)(update(evt2)) ++ Seq.fill(4)(update(evt3)))
+      processWrite(4, Seq.fill(4)(update(evt4)))
+    }
+    "write events in a single batch if the number of generated events during processing since the last write is equal to settings.writeBatchSize" in {
+      val actor = recoveredStatelessProcessor()
+
+      val evt = event("y", 1).copy(vectorTimestamp = timestamp(1, 0))
+      actor ! Written(evt)
+
+      // process first write
+      processWrite(1, Seq.fill(10)(update(evt)))
+    }
+    "stop if the number of generated events during processing of a single event is greater than settings.writeBatchSize" in {
+      val settings = new EventsourcedProcessorSettings(system.settings.config)
+      val actor = recoveredStatelessProcessor()
+
+      val evt = event("z", 1).copy(vectorTimestamp = timestamp(1, 0))
+      actor ! Written(evt)
+
+      appProbe.expectMsgType[EventsourcedProcessingResultTooLargeException].getMessage should be(new EventsourcedProcessingResultTooLargeException(11, 10).getMessage)
     }
   }
 }

